Bug 1749451: Capture sum of cpu capacity cores by OS (RHCOS or RHEL) with version, arch, and role for the nodes.

chambridge opened the pull request:
* Intended to be utilized for subscription reconciliation when cluster has mixed set of nodes (bring your own worker)

chambridge commented:
Example Output:
```
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_beta_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",node="crc-56mmj-master-0",os_image="Red Hat Enterprise Linux CoreOS 410.8.20190807.0 (Ootpa)"}
```

LiliC commented:
/ok-to-test

LiliC commented:
cc @brancz you reviewed the initial PR, do you mind having a look.

fabianofranz commented:
@smarterclayton can you help reviewing this? I'd like to move on to start testing in subscriptions.

Apart the other corner cases we've been discussing (sockets, physical, etc), this PR should give us enough tools to properly cover the golden path, the most common use cases for subscriptions.

brancz commented:
We're first going to want to have https://github.com/openshift/kube-state-metrics/pull/16 land to have a reliable source for node roles.

brancz commented:
this needs to be rebased and use the `kube_node_role` label instead of the recording rule.

chambridge commented:
@brancz rebased and swapped to use `kube_node_role`. Updated the output above as well.

@smarterclayton

chambridge commented:
/retest

s-urbaniak commented:
https://github.com/openshift/kube-state-metrics/pull/16 merged, this is green, hence we are good to go, I think :)

s-urbaniak commented:
/lgtm

brancz commented:
/hold

I don't believe the `node` dimension is supposed to be in there. That means we get the same amount of metrics as nodes in a cluster.

chambridge commented:
Happy to change it if necessary. I assumed we wanted this so that we could get the distinct cpu_capacity_cores since a node could have multiple roles.

@smarterclayton thoughts?

smarterclayton commented:
I would have expected to get the following series:

```
cpu_cores{node_role_label="blah"} 10
cpu_cores{node_role_label="blah",node_role_label="baz"} 20
cpu_cores{node_role_label="baz"} 5
```

and have there be no overlap between the groups (a node with both labels is in the middle series, a node with one label is in the last series)

chambridge commented:
If I edit a node to have an additional role, node-role.kubernetes.io/baz: '', I will get the following output from `kube_ndoe_role`:
```
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-129-18.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="master",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-130-116.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="worker",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-131-125.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="baz",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-131-125.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="worker",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-141-141.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="master",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-149-230.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="worker",service="kube-state-metrics"}	1
kube_node_role{endpoint="https-main",instance="10.128.2.6:8443",job="kube-state-metrics",namespace="openshift-monitoring",node="ip-10-0-150-171.ec2.internal",pod="kube-state-metrics-678b97f7cb-xg2mn",role="master",service="kube-state-metrics"}	1
```

I since you get a new time series per role I'm not sure how you would combine this per @smarterclayton 's comments above.


This did find a problem my existing I need to swap kube_node_role and kube_node_label due to cardinality (fixed below with order swap):
````
sum(kube_node_role * on(node) group_left(label_kubernetes_io_arch, label_node_openshift_io_os_id) kube_node_labels * on(node) group_left(os_image, kernel_version) kube_node_info * on(node) group_left kube_node_status_capacity_cpu_cores) BY (node, kernel_version, os_image, label_node_openshift_io_os_id, label_beta_kubernetes_io_arch, role)
```
output is:
```
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-141-141.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="master"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-149-230.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="worker"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-150-171.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="master"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-129-18.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="master"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-130-116.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="worker"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-131-125.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="baz"}	4
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",node="ip-10-0-131-125.ec2.internal",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="worker"}	4
```

If I drop node from dimension you get the following:
```
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="master"}	12
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="worker"}	12
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="baz"}	4
```

If I just do `cluster:capacity_cpu_cores:sum` you will get the following:
```
cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="m4.xlarge"}	12
cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="m4.xlarge",label_node_role_kubernetes_io="master"}	12
```
So you can see the overlap created by `baz`.

Any tips @LiliC @brancz ?

chambridge commented:
If we wanted to just document not to overloap `master`, `worker`, `infra` on the same node we could do the following:
```
sum(kube_node_role{role=~"master|worker|infra"} * on(node) group_left(label_kubernetes_io_arch, label_node_openshift_io_os_id) kube_node_labels * on(node) group_left(os_image, kernel_version) kube_node_info * on(node) group_left kube_node_status_capacity_cpu_cores) BY (kernel_version, os_image, label_node_openshift_io_os_id, label_beta_kubernetes_io_arch, role)
```
and in this case dropping the node dimension gives you the calculation:
```
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",os_image="Red  Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="master"} | 12
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_node_openshift_io_os_id="rhcos",os_image="Red  Hat Enterprise Linux CoreOS 42.80.20190907.0 (Ootpa)",role="worker"} | 12
```

chambridge commented:
Here is a potential idea I'll throw out:

Essentially to focus on the tags we currently care about `master`, `infra`, worker` we could create/augment some relabeling rules, building on `cluster:master_nodes`:
```
{
expr: 'kube_node_labels and on(node) kube_node_role{role="master"}',
labels: {
    label_node_role_kubernetes_io_master: 'true',
},
record: 'cluster:master_nodes',
},
{
expr: 'kube_node_labels and on(node) kube_node_role{role="infra"}',
labels: {
    label_node_role_kubernetes_io_infra: 'true',
},
record: 'cluster:infra_nodes',
},
{
expr: 'kube_node_labels and on(node) kube_node_role{role="worker"}',
labels: {
    label_node_role_kubernetes_io_worker: 'true',
},
record: 'cluster:worker_nodes',
},
```
Now we have separated node role we could do something like the following to combine it:
```
{
expr: 'sum((cluster:master_nodes * on(node) group_left(os_image, kernel_version) kube_node_info * on(node) group_left kube_node_status_capacity_cpu_cores) or on(node) (cluster:infra_nodes * on(node) group_left(os_image, kernel_version) kube_node_info * on(node) group_left kube_node_status_capacity_cpu_cores) or on(node) (cluster:worker_nodes * on(node) group_left(os_image, kernel_version) kube_node_info * on(node) group_left kube_node_status_capacity_cpu_cores)) BY (kernel_version, os_image, label_node_openshift_io_os_id, label_beta_kubernetes_io_arch, label_node_role_kubernetes_io_master, label_node_role_kubernetes_io_infra, label_node_role_kubernetes_io_worker)',
record: 'node_role_os_version_machine:cpu_capacity_cores:sum',
},
```
and the output would look like this:
```
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_beta_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190910.0 (Ootpa)"}	12
{kernel_version="4.18.0-80.7.2.el8_0.x86_64",label_beta_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_worker="true",os_image="Red Hat Enterprise Linux CoreOS 42.80.20190910.0 (Ootpa)"}	12
```

Thoughts? @LiliC @brancz @smarterclayton

smarterclayton commented:
Note that `os_image` is not the right value.  We have a label that is `rhel|rhcos|<other>` and then another one that is `7|8|<other>` and we need both of those.  We don't want the strings, we need the labels set by machine-config.

chambridge commented:
Ok. I have rebased and updated the rules.

I tweaked `cluster:master_nodes` to leave the old label `label_node_role_kubernetes_io: 'master',` since there appeared to be several other items keying of this that can be cleaned up in another PR.

I added `cluster:infra_nodes` and `cluster:worker_nodes`.

I dropped `os_image` and `kernel_version` per Clayton's comments above.
Clayton and I chased down the fact that `node.openshift.io/os_version` was removed with https://github.com/openshift/machine-config-operator/pull/657
Another issue, https://github.com/openshift/machine-config-operator/issues/582, looks to add support for rhel major version. When this is introduced we can add `label_node_openshift_io_rhelmajor` to they BY clause.

Currently the output looks like this:
```
{label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true"}	12
{label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra="true"}	4
{label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_worker="true"}	8
```

LiliC commented:
> I dropped os_image and kernel_version per Clayton's comments above.

Not sure which comments you are referring to? But the `kube_node_info` os_image label data comes from OS Image reported by the node from `/etc/os-release`. And if I understand by the title of this PR (or rather Bugzilla) we want to include the version as well:

> Capture sum of cpu capacity cores by OS (RHCOS or RHEL) with version, arch, and role for the nodes.

> I tweaked cluster:master_nodes to leave the old label label_node_role_kubernetes_io: 'master', since there appeared to be several other items keying of this that can be cleaned up in another PR.

You mean other rules relly on this?

chambridge commented:
@LiliC Clayton's original comment was:
> Note that os_image is not the right value. We have a label that is rhel|rhcos|<other> and then another one that is 7|8|<other> and we need both of those. We don't want the strings, we need the labels set by machine-config.

This is why I mentioned getting `label_node_openshift_io_rhelmajor` once openshift/machine-config-operator#582 is done. If you think it is necessary to add back `os_image` in the meantime that is fine.

The following rules are using `label_node_role_kubernetes_io`
- `cluster:capacity_cpu_cores:sum`
- `cluster:capacity_memory_bytes:sum`
- `cluster:node_instance_type_count:sum`

I also find this in some grafana dashboards and metrics.go

LiliC commented:
/test e2e-aws-upgrade

LiliC commented:
> If you think it is necessary to add back os_image in the meantime that is fine.

Nope, all good. Just wanted to double check, I saw the conversation on node slack channel now as well.

chambridge commented:
@LiliC Sounds good and thanks 🙇  for kicking off the failing test. Looks like all test now pass.

smarterclayton commented:
> I added cluster:infra_nodes and cluster:worker_node

Why?  We can join on node roles for those queries - why add those?

smarterclayton commented on jsonnet/rules.jsonnet:
```diff
@@ -45,12 +45,27 @@
             },
             record: 'cluster:master_nodes',
           },
+          {
```
Remove these, I don't see a need

smarterclayton commented on jsonnet/rules.jsonnet:
```diff
@@ -91,6 +106,10 @@
             record: 'instance:etcd_object_counts:sum',
           },
+          {
+            expr: 'sum((cluster:master_nodes * on(node) group_left kube_node_status_capacity_cpu_cores) or on(node) (cluster:infra_nodes * on(node) group_left kube_node_status_capacity_cpu_cores) or on(node) (cluster:worker_nodes * on(node) group_left kube_node_status_capacity_cpu_cores)) BY (label_node_openshift_io_os_id, label_kubernetes_io_arch, label_node_role_kubernetes_io_master, label_node ...
```
Remove worker, that's an irrelevant label.  worker role has no meaning for us.

chambridge replied:
ok

smarterclayton commented on assets/prometheus-k8s/rules.yaml:
```diff
@@ -291,6 +300,12 @@ spec:
     - expr: sum(etcd_object_counts) BY (instance)
       record: instance:etcd_object_counts:sum
+    - expr: sum((cluster:master_nodes * on(node) group_left kube_node_status_capacity_cpu_cores)
+        or on(node) (cluster:infra_nodes * on(node) group_left kube_node_status_capacity_cpu_cores)
```
So this won't tell me that it's both a master and infra node?

chambridge replied:
So if a node is marked `master` and `infra` it would be bucket into `master` because BY condition and its order.

chambridge commented:
@smarterclayton I added the additional relabeling metrics to handle the issue described here:
https://github.com/openshift/cluster-monitoring-operator/pull/466#issuecomment-529660078
and discussed on the forum-monitoring thread.

chambridge commented:
@smarterclayton I'm not clear on next steps here. I can remove the `cluster:worker_nodes` relabel metric and remove that portion from the `node_role_os_version_machine:cpu_capacity_cores:sum` rule:
```
          {
            expr: 'sum((cluster:master_nodes * on(node) group_left kube_node_status_capacity_cpu_cores) or on(node) (cluster:infra_nodes * on(node) group_left kube_node_status_capacity_cpu_cores)) BY (label_node_openshift_io_os_id, label_kubernetes_io_arch, label_node_role_kubernetes_io_master, label_node_role_kubernetes_io_infra)',
            record: 'node_role_os_version_machine:cpu_capacity_cores:sum',
          },
```
and the resulting output would be:
```
{label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_master="true"}	12
{label_kubernetes_io_arch="amd64",label_node_openshift_io_os_id="rhcos",label_node_role_kubernetes_io_infra="true"}	4
```

You could then reference `cluster:capacity_cpu_cores` to get the total, if needed.
```
cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="m4.xlarge"}	12
cluster:capacity_cpu_cores:sum{label_beta_kubernetes_io_instance_type="m4.xlarge",label_node_role_kubernetes_io="master"}	12
```

However, as you pointed out you will not see both `master` and `infra` on the same line.
